ENH: add lens calculations and configuration

cristinasewell opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Multiple calculations functions have been added (imported) from the old code to the `be_lens_calcs` module.
These functions are used for the Beryllium Lens Calculations. 
<!--- Describe your changes in detail -->
For the most part these calculations are the same as the old calculations, but there were a few changes: 
1. `get_delta()`  and `get_att_len()` functions has been changed to use the `xraydb` package instead of `periodictable` - this changed the results a bit, but the new values should be closer to the CXRO reference, so this should be an improvement from the old calculations.
2. We added a  finite source distance for planning lens sets. Previously the calculation assumed infinite source distance which in practice meant people usually had to move the beamline lenses upstream to get the focus where they wanted it - so `calc_beam_fwhm` now has an extra attribute `source_distanc` and some logic to handle that if provided. 
3. A new function has been added `configure_lens_set_file` that will allow the user to configure the `lens_set_file`  - this was done differently in the old python code. 
4. The `get_lens_set` and `set_lens_set` have been changed from the old code. Now we are using `numpy.save` and `numpy.load` to store and get the lens sets - using a binary Numpy `.npy` file format.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Like Zach and Ken said: ***It is appropriate to provide stand-alone calcs as well, separate from the BeLens class. There is some use for sitting down at the terminal and running a calculation without thinking about or interfacing with specific device hardware.***
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
The functions have been tested against the old functions and they gave the same results. After confirming that the values are as expected, `get_delta()` and `get_att_lenght()` have been changed so the numbers changed too, and these numbers have been checked and approved by the scientists. :) 
Also the FWHM_SIGMA_CONVERSION coefficient has been changed to use more precision = 2.35482004503 - this changed the numbers a bit...

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Where Has This Been Documented?
<!--  Include where the changes made have been documented. -->
<!--  This can simply be  a comment in the code or updating a docstring -->

<!--
## Screenshots (if appropriate):
-->

cristinasewell committed:
WIP - add some old calculations for reference

mseaberg commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,360 @@
+        print("size           : %.3e" % (size))
+        print("size FWHM      : %.3e" % (size_fwhm))
+
+    return size_fwhm
```
Adding source_distance argument and a couple lines to modify distance to focus if this is provided. We could also make defaults for each hutch.

```suggestion
 def calc_beam_fwhm(self, E, lens_set, distance=None, source_distance=None, material="Be",
                       density=None, fwhm_unfocused=None, printsummary=True):
    """
    Calculates beam parameters for certain lenses configuration
    and energy at a given distance.
    Optionally some other parameters can be set
    Beamsize for certain lenses configuration and energy at a given distance

    Parameters
    -----------
    E: number
        Beam Energy
    lens_set : `list`
        [(numer1, lensthick1), (number2, lensthick2)...]
    distance: float
        Distance from the lenses to the sample is 3.852 m at XPP.
    source_distance: float
        Distance from source to lenses. This is about 160 m at XPP.
    material: str
        Beryllium. The use of beryllium extends the range of operation
        of compound refractive lenses, improving transmission,
        aperture size, and gain
    density: TODO: find out what density is
    fwhm_unfocused: float
        This is about 400 microns at XPP.
    printsummary: boolean
        Prints summary of parameters/calculations if True

    Returns
    -------
    float: Size FWHM
    """
    # Focal length for certain lenses configuration and energy
    focal_length = calc_focal_length(E, lens_set, material, density)
    
    # use lens makers equation to find distance to image of source
    if source_distance is not None:
        focal_length = 1 / (1 / focal_length - 1 / source_distance)
    
    lam = 1.2398 / E * 1e-9
    # the w parameter used in the usual formula is 2*sigma
    w_unfocused = fwhm_unfocused * 2 / 2.35
    # assuming gaussian beam divergence = w_unfocused/f we can obtain
    waist = lam / np.pi * focal_length / w_unfocused
    rayleigh_range = np.pi * waist ** 2 / lam
    size = waist * np.sqrt(1.0 + (distance - focal_length) ** 2.0 / rayleigh_range ** 2)
    size_fwhm = size * 2.35 / 2.0

    if printsummary:
        print("FWHM at lens   : %.3e" % (fwhm_unfocused))
        print("waist          : %.3e" % (waist))
        print("waist FWHM     : %.3e" % (waist * 2.35 / 2.0))
        print("rayleigh_range : %.3e" % (rayleigh_range))
        print("focal length   : %.3e" % (focal_length))
        print("size           : %.3e" % (size))
        print("size FWHM      : %.3e" % (size_fwhm))

    return size_fwhm
```

cristinasewell committed:
Add source_distance to calc_beam_fwhm

Adding source_distance argument and a couple lines to modify distance to focus if this is provided

Co-authored-by: Matt Seaberg <54456229+mseaberg@users.noreply.github.com>

cristinasewell committed:
WIP - a bit more cleanup

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,368 @@
+    # if type(lens_set) is int:
+    #     lens_set = getLensSet(lens_set)
+    for i in range(len(lens_set) / 2):
+        # TODO: why new code is (double //) range(len(lens_set)//2):
```
`//` means floor division - https://docs.python.org/3/glossary.html#term-floor-division

It's appropriate here as Python 3 `range` will not take a `float`.

For fun, an alternative for this loop could be using numpy to reshape it:
```
In [7]: np.arange(20).reshape((10, 2))[:, 0]
Out[7]: array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [8]: np.arange(20).reshape((10, 2))[:, 1]
Out[8]: array([ 1,  3,  5,  7,  9, 11, 13, 15, 17, 19])

# then zip(_7, _8)
```

Considering it looks like there may be `None` in there, using Python to `zip` items together may be preferable:
```python
In [9]: items = np.arange(20)

In [10]: print(list(zip(items[::2], items[1::2])))
[(0, 1), (2, 3), (4, 5), (6, 7), (8, 9), (10, 11), (12, 13), (14, 15), (16, 17), (18, 19)]
```

cristinasewell replied:
Great, thank you, I added the `//` back yesterday after reading about it (well after getting the error that we can't use range for float haha) - that was something I did not know we can do, but I like the the idea of using numpy - i'll give it a try. 

Also, there are a few places I noticed with possible having `None` - I will try this too - thanks a lot

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,368 @@
+    E: number
+        Beam Energy
+    lens_set : `list`
+        [(numer1, lensthick1), (number2, lensthick2)...]
```
According to the code it's:
```suggestion
        [numer1, lensthick1, number2, lensthick2...]
```

cristinasewell replied:
yes that's right, I realized that after some testing :) - thank you @klauer

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,368 @@
+    size_fwhm = size * 2.35 / 2.0
+
+    if printsummary:
+        print("FWHM at lens   : %.3e" % (fwhm_unfocused))
```
For stuff like this that'll likely be repeated in many spots, I think we should provide recommended ways of doing things. Perhaps that could be something like using [tabulate](https://pypi.org/project/tabulate/) to create tables?

cristinasewell replied:
I'll give that a try, I think that's a great idea, thank you

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,368 @@
+            print("somehow failed ...")
+            break
+        absdiff = abs(distance - f)
+    print("Energy that would focus at a distance of %.3f is %.3f" % (distance, E))
```
We should also try to enforce `print` -> `logger` changes, I think.

cristinasewell replied:
Ok that was something I was not sure about, i'll switch to using `logger`

klauer reviewed (commented):
Random thoughts that don't necessarily belong in this PR:

We have a bit of an opportunity to shape what goes in this repository.

That doesn't necessarily mean writing calculation routines for scientists (that's not so likely) - but standardizing certain aspects such as printing/logging/documentation/layout/etc. As those that review things that roll into this repository, we can try to better guide the direction and improve things for all involved.

klauer commented:
My out-of-place review was better suited for an issue (sorry about that) - see #6.

cristinasewell committed:
WIP: add changes from Matt and a bit more cleanup

cristinasewell commented:
> My out-of-place review was better suited for an issue (sorry about that) - see #6.

Ok that sounds good, I'll still leave some things that are related to #6 there.... since I have some already, but that's good, thank you

Please feel free review - I don't mind at all :) it make it easier for me

mseaberg commented:
@cristinasewell I will sort of follow along here but probably won't plan to follow in detail. Let me know anywhere you want my input. For some of this stuff if we want it to be well-documented it's almost easier to re-write than to figure out what was going on...

cristinasewell commented:
> @cristinasewell I will sort of follow along here but probably won't plan to follow in detail. Let me know anywhere you want my input.

Thank you very much @mseaberg - I will tag your name in the comment so you get a notification.

cristinasewell committed:
WIP: a bit more cleanup

cristinasewell committed:
WIP: more cleanup

cristinasewell committed:
TST: brought some tests from test_lens but needs to be revised and more tests

cristinasewell committed:
TST: added more test valuse based on old functionality.. and some corrections

cristinasewell commented:
Should I implement these functions here too - the `setLensSetsToFile` and `getLensSet`? 
https://github.com/pcdshub/old-hutch-configs/blob/18c7769d66a647ab558ad7ebc34d019fc98fbf48/xcs/prod/utilitiesBeLenses.py#L15-L50

In the old code we were using the `getLensSet` in `calcFocalLength`
https://github.com/pcdshub/old-hutch-configs/blob/18c7769d66a647ab558ad7ebc34d019fc98fbf48/xcs/prod/utilitiesBeLenses.py#L80

cristinasewell committed:
WIP: get_delta and get_att_len with xraydb module

cristinasewell committed:
WIP: add the three functions for transmission

cristinasewell committed:
TST: test the threee transmission functions agains the old code

cristinasewell committed:
MNT: add set_lens_set_to_file and get_lens_set functions

cristinasewell committed:
TST: test the set_lens_set_to_file and get_lens_set a bit, need more testing

cristinasewell committed:
WIP: add find_z_pos and cleanup calc_lens_set

cristinasewell committed:
TST: test calc_lens_set, find_energy and find_z_pos against old code

cristinasewell committed:
WIP: using the new get_delta and get_att_len

cristinasewell committed:
TST: adjusted tests with new get_delta and get_att_len

cristinasewell committed:
MNT: remove a silly comment

cristinasewell commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,732 @@
+logger = logging.getLogger(__name__)
+
+# We have sets of Be lenses with thicknesses:
+LENS_RADII = [50e-6, 100e-6, 200e-6, 300e-6, 500e-6, 1000e-6, 1500e-6]
```
Is the `LENS_RADII` the latest one we have? - should I add `2000e-6` too?
Based on https://sites.google.com/site/xppwiki/beamline-components/be-lenses we have the following:

***We have sets of Be lenses with thicknesses: 50 um (rarely used because small aperture, only for really hard x-rays), 100 um, 200 um, 300 um, 500 um, 1000 um, 1500 um and 2000 um.***

ZryletTC replied:
This matches what's in xpp and xcs's old python, but I wouldn't know what's currently being used.
I suggest asking Eric, Diling, and/or Matt

cristinasewell commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,732 @@
+LENS_RADII = [50e-6, 100e-6, 200e-6, 300e-6, 500e-6, 1000e-6, 1500e-6]
+
+
+def get_lens_set(set_number_top_to_bot, filename):
```
Do we need to provide a specific path here, maybe current directory or somewhere where we save these? or the user will deal with that?

cristinasewell commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,732 @@
+    return energy
+
+
+def find_z_pos(energy, lens_set, spot_size_fwhm, material="Be",
```
we might now need the  `find_z_pos` here? - had this in the old code

cristinasewell commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,732 @@
+    return radius
+
+
+def find_energy(lens_set, distance=3.952, material="Be", density=None):
```
we might not need the `find_energy` ? - this was part of the old code too

ZryletTC replied:
Might as well leave it in for now.

cristinasewell commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,732 @@
+
+# TODO: ========== WE MIGHT NOT NEED THESE FUNCTIONS BELOW =================
+# TODO: distance default might need to be changed
+def find_radius(energy, distance=4.0, material="Be", density=None):
```
might not need `find_radius` here?  - we had it in the old code

klauer replied:
Can leave it in for now, I suppose, since you've already done the work here.

cristinasewell commented on pcdscalc/tests/test_be_lens_calcs.py:
```diff
@@ -0,0 +1,299 @@
+])
+def test_att_length(energy_sample, expected):
+    # old values (with old get_att_lenght)
+    # 0.005949593420831057),
```
the values used for the new implementation of the `get_att_len` seem pretty far off :(  - but I am not sure if we were using before the same `getAttLen` as I used...

cristinasewell commented:
I have a few questions: 

 

1.  I am using some Constants in the calculations (that I don't know what they are yet):
```
1.2398
2.35
12.398
2.35482 
```
But should we have One location where we keep all these constants and not use raw numbers??

2.  Also, there are lots of defaults that we are using in the functions, something like this for example:
```
calc_lens_set(energy, size_fwhm, distance, n_max=12, max_each=5,
                  lens_radii=[100e-6, 200e-6, 300e-6, 500e-6, 1000e-6],
                  fwhm_unfocused=0.0005, eff_rad0=None):
```
Should I use some type of constants here too? what if the defaults will change? and also, the defaults probably already changed so I probably should go back and make sure all the defaults are valid. 

3. One more question/concern: Testing these calculations is a bit tricky, as I don't know what really the results should be. 
What I did is I first tested all the functions against the old code - and I got the same results. The problem came where I changed the implementation of `get_att_len` and `get_delta` - because multiple functions use these two, I got different results, so what I did is I brought the new `get_att_len` and `get_delta` in the OLD code and I tested my functions again... but this is not a very good practice - any suggestions please?

cristinasewell commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,732 @@
+    # TODO: might need these 2 lines here
+    # or can we do something else here to
+    # know when we want to get the len_set from the file?
+    # if type(lens_set) is int:
```
I am adding the ability to get the `lens_set` from the file here. The user will just provide the number of the set in the file, but we also need a `filename` - will we have a set path for the file here?

mseaberg commented:
> I have a few questions:
> 
> 1. I am using some Constants in the calculations (that I don't know what they are yet):
> 
> ```
> 1.2398
> 2.35
> 12.398
> 2.35482 
> ```
> 
> But should we have One location where we keep all these constants and not use raw numbers??
> 
> 2. Also, there are lots of defaults that we are using in the functions, something like this for example:
> 
> ```
> calc_lens_set(energy, size_fwhm, distance, n_max=12, max_each=5,
>                   lens_radii=[100e-6, 200e-6, 300e-6, 500e-6, 1000e-6],
>                   fwhm_unfocused=0.0005, eff_rad0=None):
> ```
> 
> Should I use some type of constants here too? what if the defaults will change? and also, the defaults probably already changed so I probably should go back and make sure all the defaults are valid.
> 
> 3. One more question/concern: Testing these calculations is a bit tricky, as I don't know what really the results should be.
>    What I did is I first tested all the functions against the old code - and I got the same results. The problem came where I changed the implementation of `get_att_len` and `get_delta` - because multiple functions use these two, I got different results, so what I did is I brought the new `get_att_len` and `get_delta` in the OLD code and I tested my functions again... but this is not a very good practice - any suggestions please?


1. I can tell you what the constants are, but won't offer an opinion on where to keep them. The 1.2398 and 12.398 have the same use, they are for converting between wavelength and photon energy. 1.2398 helps to convert between keV and nm:
`λ[nm] = 1.2398/E[keV]`
 and 12.398 would convert between keV and angstroms (sometimes the units for these index of refraction tables are in angstroms). Probably the best practice would be to implement a function that converts from photon energy to wavelength. The 2.35 and 2.35482 are also the same thing, just to different levels of precision. This is for converting between FWHM and sigma of a Gaussian function:
`FWHM=2.355*sigma`

2. The defaults were probably set by preferences at XPP. They could be revisited or at least the justification can be provided in the docstring.

3. For testing, if you paste in the results you got after changing `get_atten_len` and `get_delta`, as well as those from the OLD code with original implementations, Eric or I can tell you whether new results look good.

cristinasewell committed:
WIP: add logic back to use file for lens_set - needs more work

cristinasewell commented:
@mseaberg thanks for the explanation of those Constants and defaults. I will paste the old results vs new results shortly - thanks a lot.

cristinasewell commented:
[calcs.xlsx](https://github.com/pcdshub/pcdscalc/files/5358106/calcs.xlsx)


**test_get_att_length** |   |  
-- | -- | --
OLD VALUES: | NEW VALUES: | TESTS:
0.005949593 | 0.48101132541206565 | get_att_len(energy=8, material='Be')
0.008747658 | 0.650740965 | get_att_lenLen(energy=9, material='Be')
9.03E-06 | 0.000894698 | get_att_lenLen(energy=1, material='Be')
0.471176114 | 3.016074691 | get_att_lenLen(energy=30, material='Be')
0.012355158 | 0.8360110829541422 | get_att_lenLen(energy=10, material='Be')
**test_get_delta** |   |  
OLD VALUES: | NEW VALUES: | TESTS:
5.33E-06 | 5.330471261281744e-06 | get_delta(8, material='Be')
4.21E-06 | 4.21E-06 | get_delta(9, material='Be')
0.000352431 | 0.0003515107795702183 | get_delta(1, material='Be')
0.003807347 | 0.005621923 | get_delta(0.1, material='Be')
8.51E-07 | 8.52E-07 | get_delta(20, material='Be')
**test_calc_focal_length_for_single_lens** |   |  
OLD VALUES: | NEW VALUES: | TESTS:
281.6132124 | 281.40101061895905 | calc_focal_length_for_single_lens(8, 0.003,   material='Be')
9.387107081546501 | 9.380033687 | calc_focal_length_for_single_lens(8, 0.0001,   material='Be')
356.5002989 | 356.2233962102125 | calc_focal_length_for_single_lens(9, 0.003,   material='Be')
0.141871555 | 0.142243148 | calc_focal_length_for_single_lens(1, 0.0001,   material='Be')
**test_calc_focal_length** |   |  
OLD VALUES: | NEW VALUES: | TESTS:
69.45047645 | 69.39814397 | calc_focal_length(8, [1, 0.02, 5, 0.004, 2, 1.23, 1, 0.02],   material="Be")
5.2150594897480556 | 5.211129826 | calc_focal_length(8, [2, 200e-6, 4, 500e-6],   material="Be")
87.91887070181892 | 87.85058192 | calc_focal_length(9, [1, 0.02, 5, 0.004, 2, 1.23, 1, 0.02],   material="Be")
1.0496361635424505 | 1.052385399 | calc_focal_length(1, [1, 0.02, 5, 0.004, 2, 1.23, 1, 0.02],   material="Be")
**test_calc_beam_fwhm** |   |  
OLD VALUES: | NEW VALUES: | TESTS:
0.000753947 | 0.000753912 | calc_beam_fwhm(energy=8,   lens_set= [1, 0.02, 5, 0.004, 2, 1.23, 1, 0.02],      distance=4,  fwhm_unfocused=800e-6, material='Be')
0.000186393 | 0.00018593 | calc_beam_fwhm(energy=8,   lens_set= [2, 200e-6, 4, 500e-6],      distance=4,  fwhm_unfocused=800e-6, material='Be')
0.000212373 | 0.000212156 | calc_beam_fwhm(energy=8,   lens_set=[2, 200e-6, 4, 500e-6],    distance=3,    fwhm_unfocused=500e-6, material='Be')
0.000471297 | 0.000471276 | calc_beam_fwhm(energy=8,   lens_set= [1, 0.02, 5, 0.004, 2, 1.23, 1, 0.02],      distance=4,  fwhm_unfocused=500e-6, material='Be')
0.000929067 | 0.000925334 | calc_beam_fwhm(energy=1,   lens_set= [1, 0.02, 5, 0.004, 2, 1.23, 1, 0.02],      distance=3,  fwhm_unfocused=500e-6, material='Be')
**test_calc_distance_for_size** |   |  
OLD VALUES: | NEW VALUES: | TESTS:
[4.00301939, 134.89793351] | [4.0, 134.79628794] | calc_distance_for_size(size_fwhm=0.0007539124064936633,   lens_set=[1, 0.02, 5, 0.004, 2, 1.23, 1, 0.02],   energy=8,  fwhm_unfocused=800e-6)
[4.00301637, 6.42710261] | [4.0, 6.42225965] | calc_distance_for_size(size_fwhm=0.00018593023997294836,   lens_set=[2, 200e-6, 4, 500e-6],   energy=8,  fwhm_unfocused=800e-6)
[3.00226229, 7.42785669] | [3.0, 7.42225965] | calc_distance_for_size(size_fwhm=0.0002121557393404887,   lens_set=[2, 200e-6, 4, 500e-6],   energy=8,  fwhm_unfocused=500e-6)
[4.00303623, 134.89791667] | [4.0, 134.79628794] | calc_distance_for_size(size_fwhm=0.00047127559879883534,   lens_set=[1, 0.02, 5, 0.004, 2, 1.23, 1, 0.02],   energy=8,  fwhm_unfocused=500e-6)
[-0.89289053,  2.99216285] | [-0.8952292, 3.0] | calc_distance_for_size(size_fwhm=0.0009253340546070372,   lens_set=[1, 0.02, 5, 0.004, 2, 1.23, 1, 0.02],   energy=1,  fwhm_unfocused=500e-6)
**test_calc_trans_for_single_lens** |   |  
OLD VALUES: | NEW VALUES: | TESTS:
0.91927896 | 0.963284221 | calc_trans_for_single_lens(energy=8, radius=1.0e-3,   material="Be",    fwhm_unfocused=900e-6, disk_thickness=1.0e-3, apex_distance=30e-6)
0.970049683 | 0.998330828 | calc_trans_for_single_len(8, 2.0e-3,   material="Be",    fwhm_unfocused=900e-6, disk_thickness=1.0e-3, apex_distance=30e-6)
0.985562648 | 0.999757145 | calc_trans_for_single_len(9, 3.0e-3,   material="Be",    fwhm_unfocused=900e-6, disk_thickness=1.0e-3, apex_distance=30e-6)
**test_calc_trans_lens_set** |   |  
OLD VALUES: | NEW VALUES: | TESTS:
0.204979658 | 0.28135231 | calc_trans_lens_set(energy=8,   lens_se=[2, 100e-6, 4, 200e-6], material="Be",   fwhm_unfocused=900e-6, disk_thickness=1.0e-3, apex_distance=30e-6)
0.36759864 | 0.483493536 | calc_trans_lens_set(energy=8,    lens_set= [2, 200e-6, 4, 500e-6],   material="Be",   fwhm_unfocused=900e-6, disk_thickness=1.0e-3, apex_distance=30e-6)
0.400546189 | 0.483947927 | calc_trans_lens_set(energy=9,    lens_se=[2, 200e-6, 4, 500e-6]. material="Be",   fwhm_unfocused=900e-6, disk_thickness=1.0e-3, apex_distance=30e-6)

mseaberg commented:
Thanks for posting the results @cristinasewell! Everything looks great except for the tests related to calculating attenuation through the lens. Since Eric suggested the xraydb package and he has already made his own comparisons maybe he can help with that.

cristinasewell commented:
> Thanks for posting the results @cristinasewell! Everything looks great except for the tests related to calculating attenuation through the lens. Since Eric suggested the xraydb package and he has already made his own comparisons maybe he can help with that.

Thank you @mseaberg - I am now thinking that maybe I am using the wrong old function for `get_atten_len`, I'll ask Eric, I can't tag him here. Thanks for checking the results :)

cristinasewell committed:
MNT: a bit more cleanup

cristinasewell committed:
MNT: correct some wording

cristinasewell committed:
MNT: remove unnecessary things and rewording

cristinasewell committed:
MNT: correct some docstring format and add usage to more functions

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,830 @@
+# Set of Be lenses with thicknesses:
+LENS_RADII = [50e-6, 100e-6, 200e-6, 300e-6, 500e-6, 1000e-6, 1500e-6]
+
+FWHM_SIGMA_CONVERSION = 2.35482004503
```
Should at least be a one-liner comment for these

cristinasewell replied:
Ok added comments for these in 7820456 - thank you.

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,830 @@
+
+def photon_to_wavelength(energy):
+    '''
+    Find the wavelength in micrometers, using the
```
Docstrings look good, but oddly formatted. Try setting your editor to wrap at 79 chars and reformat the block as appropriate.

In Vim, for example, can do this easily when you do a visual selection of these lines and type `gw` in normal mode. I'm sure if you're using an IDE there's something similar.

cristinasewell replied:
I ended up using Vim to correct some of the docstrings  (thanks for letting me know how), because my IDE (Visual Studio Code) would only show what word I should wrap at, but not do it automatically - I need to investigate my IDE a bit more though to see if I have this option. 

I still don't know for sure if the docstrings look fine now - please let me know.

klauer replied:
Much better!

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,830 @@
+
+    Parameters
+    ----------
+    fwhm : `float`
```
```suggestion
    fwhm : float
```

More details in the style guide - it'd be good to become more familiarized with it:
https://pcdshub.github.io/style.html

cristinasewell replied:
thank you - yes I should

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,830 @@
+
+def set_lens_set_to_file(sets_list_of_tuples, filename, make_backup=True):
+    '''
+    # The Be lens holders can take 3 different sets that we usually set before
```
Comments in docstring?

cristinasewell replied:
oops hahah - took it out

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,830 @@
+    with open(filename, 'w') as lens_file:
+        # TODO: do we want to pprint.pformat() like in the old code?
+        # lens_file.write(pprint.pformat(sets_list_of_tuples))
+        lens_file.write(sets_list_of_tuples)
```
It's not a great idea to write stringified lists of python values to a file and `eval()` their contents on the way out. If the scientists aren't requiring this file format, this could be standardized to just use numpy's built-in file saving/loading utilities (https://numpy.org/doc/stable/reference/generated/numpy.save.html).

cristinasewell replied:
ok I change to using numpy save/load in 965a3e8, please let me know if that's what you meant, as I have never used this one before :) 
Thanks for the suggestion

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,830 @@
+
+    Returns
+    -------
+    att_len : `numpy.float64`
```
```suggestion
    att_len : float
```

No need to be specific about this numpy type here - it functions just as a regular Python float.

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,830 @@
+                                         density=density)) * 1.0e-2
+
+    except Exception as ex:
+        logger.error('Get Attenuation Lenght error: %s', ex)
```
```suggestion
        logger.error('Get Attenuation Length error: %s', ex)
```

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,830 @@
+    # TODO: how can i fix this hack here:
+    # this is an ugly hack: the radius will never be bigger than 1m,
+    # so will always be overwritten
+    radius_aperture = 1.0
```
Not so ugly a hack as the above comment indicates. "Better" option might be to build and work with well-named lists (e.g., `apertures = [... for num, radius in lens_set]; radius_aperture = min(apertures)`)

cristinasewell replied:
Ok I want to make sure I understand this one. I would have something like this: ??
```
 if isinstance(lens_set, int):
        lens_set = get_lens_set(lens_set)
    apertures = list(lens_set[1::2])
    radius_aperture = min(apertures)
```
  don't I need `radius_aperture = max(apertures)`?

I might not understand exactly what this is doing....

klauer replied:
It's hard to further explain without just showing it in code:

```python
lens_set = list(zip(lens_set[::2], lens_set[1::2]))
radius_total_inv = sum(num / radius for num, radius in lens_set)
apex_distance_tot = sum(num * apex_distance for num, _ in lens_set)
radius_aperture = min(np.sqrt(radius * (disk_thickness - apex_distance))
                      for _, radius in lens_set)
```

(It's very possible I messed up something above, so don't take it verbatim - it's just an example of how to remove the for-loop and arbitrary starting value for `min(...)`)

cristinasewell replied:
ohhh got it, thanks so much - sorry I did not realize that we're trying to do. 🤦 
I was trying to:
```
radius_aperture = max(apertures)
```
and then in the loop again: 
```
radius_aperture = min(radius_aperture, new_rad_ap)
```

Thanks a lot!!

klauer replied:
Full disclosure: I didn't look at the original code, just what's in the PR. Please do double/triple check it.

cristinasewell replied:
I changed this to use your suggested code in 33e8f54, and passed the tests, I am getting the same results. 
I am pretty convinced that it works the same in the old code, I implemented these changes in the old code too and get the same results. 
Thanks a lot

klauer replied:
Great 👍

klauer commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,830 @@
+            logger.error("somehow failed ...")
+            break
+        abs_diff = abs(distance - focal_length)
+    logger.info("Energy that would focus at a distance of %.3f is %.3f"
```
```suggestion
    logger.info("Energy that would focus at a distance of %.3f is %.3f", distance, energy)
```

cristinasewell replied:
fixed it

klauer reviewed (commented):
Looking good overall 👍 

As to the question of "how can we explain where to find this stuff" - we have Sphinx-based generated documentation in place for this repo. I'd recommend adding this modules to the documentation along with any additional notes you might have there. Docs could either be in this PR or a follow-up one.

cristinasewell committed:
STY: formated the docstring to follow the PCDS Python Style Guide a bit more

cristinasewell committed:
STY: corect a few more pep8 things

cristinasewell committed:
MNT: use numpy to save/load lens_sets

cristinasewell committed:
MNT: use suggestion from Ken for calc_trans_lens_set to get rid of 'ugly hack'

cristinasewell commented:
> As to the question of "how can we explain where to find this stuff" - we have Sphinx-based generated documentation in place for this repo. I'd recommend adding this modules to the documentation along with any additional notes you might have there. Docs could either be in this PR or a follow-up one.

Ok I'll add it in this PR - thank you

ZryletTC commented on requirements.txt:
```diff
@@ -2,3 +2,4 @@ numpy
 periodictable
 scipy
 versioneer
+xraydb
```
Need to add it to `conda-recipe/meta.yaml` as well.

Also, is periodictable still needed if xraydb is being used instead?

cristinasewell replied:
Ohh thank you, I'll add it to conda-revipe too 
I am not using `periodictable` in my code, so  I should probably take it out for now. Thank you.

ZryletTC replied:
Looks like xraydb is on nsls2forge. @klauer Do we want to use that?

klauer replied:
In my opinion, we should really try to get it on conda-forge. We've had such problems with channel priorities and conda settings and building environments, I'd be a bit scared at adding nsls2forge to the mix.

cristinasewell replied:
From my lack of knowledge I don't fully understand what this means for this ticket. Will we have to create a conda-recipe for the `xraydb` or how does it work? Is conda-recipe only taking packages form `conda-forge` unless we specify otherwise?

klauer replied:
Working on getting xraydb on conda-forge separately - but it may take a while for conda-forge to accept it. Most of it is entirely out of scope for this PR.

(Ref: https://github.com/conda-forge/staged-recipes/pull/12898)

cristinasewell replied:
ohh that's great, thanks a lot @klauer!

klauer replied:
This moved much more quickly than expected. It should be available on conda-forge very soon, ready for this PR's purposes.

ZryletTC commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,826 @@
+    return fwhm / FWHM_SIGMA_CONVERSION
+
+
+# TODO: this is a test file... take care of it
```
What still needs to be done for the test file to become finalized?

cristinasewell replied:
Well in the old code it seems like we had a fixed path for this as a default, so I don't know if we should still have that here or if we should leave it up to the user, but also, if we do need a default path for it, how should I go about it exaclty?

https://github.com/pcdshub/old-hutch-configs/blob/18c7769d66a647ab558ad7ebc34d019fc98fbf48/xpp/blutil/belenses.py#L15-L21

ZLLentz replied:
We probably want a way to configure the path to the lens set file, and then we would set it up as needed in the hutch python loader. Then the default argument would be "whatever the user last configured".

The approach in the old hutch python (creating a dummy config module) should not be used, as this can cause package incompatibilities. I'd advocate for simple configuration functions in this module, e.g. `configure_lens_set_file` or some equivalent.

cristinasewell replied:
Ok @ZLLentz  I like this idea, but I am not sure if I implemented it as you intended.
I added it in 5729619 but have a couple of fixes in 0244c55 and dd34461 - please let me know if this implementation if fine.

ZLLentz replied:
Seems good to me

ZryletTC commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,826 @@
+
+    except Exception as ex:
+        logger.error('Get Attenuation Length error: %s', ex)
+        return
+    # TODO: if we return we'll get att_len == None, is this what we want?
```
```suggestion
        raise
```
I'd expect that we'd instead want to re-raise the exception. @klauer or @ZLLentz, what do you think?

klauer replied:
Agreed. Assuming the direct caller _isn't solely_ a scientist sitting at an IPython console, then catching the exceptions, emitting log messages, and returning `None` is asking for trouble down the line.

ZLLentz replied:
I agree as well, it's better to raise an exception than to change the return type. Then you can catch and react to these things in clearly readable ways in the calling code.

cristinasewell replied:
ok thanks a lot, i added this in 0f33d1e and fixed it a bit in f7d9bf1

ZryletTC commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,826 @@
+    return att_len
+
+
+# TODO: do we still need the density here? we are now
+# using density=xdb.atomic_density(material)? or should I
+# use xdb.atomic_density only if density == None?
```
> or should I use xdb.atomic_density only if density == None?

I like that idea

cristinasewell replied:
ok I can do that - wouldn't this allow for user error? - using the wrong density? Or are there situations where we want to provide densities other than the ones `xrayDB` is using?

ZLLentz replied:
It's not obvious to me how often the density argument will be used, but I think we should keep it anyway since it's not so hard to include.

ZryletTC replied:
Yeah, I don't expect anyone to use the parameter much so you don't really have to worry about user error but it's always nice to be able to override some database value, for testing/debugging, if xrayDB is missing a value of something, or maybe we have some experimental value that works better 🤷‍♂️

klauer replied:
`periodictable` had the wrong density for carbon, as it didn't assume diamond that we use in AT2L0's filters (https://github.com/pcdshub/hxr-attenuator/blob/b22dd49564e782f86665bcaefec0899e153052f4/ioc-lfe-at2l0-calc/calculator.py#L30). I'm not sure how xraydb works, but this override may be more useful than it seems.

cristinasewell replied:
Ok sounds great, thank you. Handled this in eb963b2

ZryletTC commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,826 @@
+    set_number_top_to_bot : int
+        The Be lens holders can take 3 different sets that we usually set
+        before experiments, this is to specify what number set.
+    filename : str
```
```suggestion
    filename : str, optional
```
Any parameters with default values should be denoted as optional. I think there are a couple more in this file.

cristinasewell replied:
Thanks a lot, added optional all (i think) default values in 0f33d1e

ZryletTC commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,826 @@
+        The Be lens holders can take 3 different sets that we usually set
+        before experiments, this is to specify what number set.
+    filename : str
+        File path of the lens_set file
```
If the user is meant to be able to provide a file, it would probably be useful to give more detail about said file or point them to the default/example file.

cristinasewell replied:
added some more info in a05ef43 - but I think I should be adding even more as part of the Documentation - thank you for pointing this out :)

ZryletTC commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,826 @@
+    ----------
+    energy : number
+        x-ray energy in KeV
+    material : str
+        Chemical formula (‘Fe2O3’, ‘CaMg(CO3)2’, ‘La1.9Sr0.1CuO4’)
+        Default - Beryllium.
```
Does only take chemical formulas? This would be useful to mention in the other functions instead of or in addition to the note about its benefits. This would also mean that 'Beryllium' could be misleading.
```suggestion
    material : str
        Chemical formula (‘Fe2O3’, ‘CaMg(CO3)2’, ‘La1.9Sr0.1CuO4’)
        Defaults to 'Be'.
```
If not, doesn't matter as much but it would still be nice to know what formats it accepts.

ZryletTC replied:
I also noticed your quotes here are actual quotes `‘` instead of apostrophes `'`. I'm guessing these are straight from https://xraypy.github.io/XrayDB/python.html#xraydb.xray_delta_beta and it should be fine in this context since it's in the docstring but I just thought it was interesting 🤔

cristinasewell replied:
So I am actually wrong here, thanks for pointing this out - it does not take a chemical formula, it takes an atomic symbol for element - i'll correct this. And thanks for the suggestion about `Be`

cristinasewell replied:
fixed it in 0f33d1e

ZryletTC commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,826 @@
+
+    Usage:
+
+    .. code-block:: python
```
Usage is helpful but I think with numpydoc standards, it's better to replace this with an Examples block

ZryletTC reviewed (commented):
Looking good!
Here's a few comments/suggestions for now

cristinasewell committed:
STY: use Examples instead of Usage

cristinasewell committed:
MNT: add optional to default values, also raise exeptions

cristinasewell committed:
DOC: add be_lens_module to documentation

cristinasewell committed:
MNT: remove periodictable from req, handle density=None

cristinasewell committed:
FIX: fix raise exception

cristinasewell committed:
MNT: add configuration function for lens_set file

cristinasewell committed:
FIX: add global LENS_SET_FILE

cristinasewell committed:
FIX: fix the use of the global LENS_SET_FILE

cristinasewell committed:
MNT: added some more comments about the .npy file format

cristinasewell committed:
TST: add a bit more testing

cristinasewell committed:
DOC: add a bit more documentation - might need some revision

cristinasewell committed:
MNT: adjusted some default values

cristinasewell commented:
I am not sure if I answered this request in this PR about the `planSet` - I thought this will be related to the Lens Classes so I did not do much here except adding the `calc_lens_set` function. 
https://github.com/pcdshub/pcdscalc/issues/1#issuecomment-704620672

Also @ZLLentz  how do we want to configure the parameters mentioned in https://github.com/pcdshub/pcdscalc/issues/1#issuecomment-709475768 - kind of the same idea that we did for the file path? Should I have a function where we can configure any parameters and just use them as constants in the function instead?

ZLLentz commented:
Yes, we probably want to wrap them all into one function if there are more than two. Then in the user application we figure out what hutch we're in and do something like
```
be_lens_calculations.configure_constants(fwhm=400, lens_radii=[<something>], ...)
```
prior to the first interactive call to any of these calculations.

If we end up with clearly defined constants for each hutch later then we could add a function here in a different PR like `configure_for_hutch('hutch_name')`

cristinasewell committed:
MNT: add configure_defaults function

cristinasewell commented:
> Yes, we probably want to wrap them all into one function if there are more than two. Then in the user application we figure out what hutch we're in and do something like
> 
> ```
> be_lens_calculations.configure_constants(fwhm=400, lens_radii=[<something>], ...)
> ```
> 
> prior to the first interactive call to any of these calculations.
> 
> If we end up with clearly defined constants for each hutch later then we could add a function here in a different PR like `configure_for_hutch('hutch_name')`

Ok I hope this 015c274 is kind of what we want?
I have not added the defaults for `n_max` and  `max_each` because we only use them in one function `calc_lens_set` - but please let me know if I should add them too.

cristinasewell committed:
DOC: add lens_radii to docstring in configure_defaults

cristinasewell committed:
FIX: add distance to configure_defaults

cristinasewell committed:
DOC: added info about configure_defaults

klauer reviewed (approved):
I skimmed over this for a final pass, and nothing major stuck out. Great work!

Before we merge, though, let's let @ZLLentz (and @ZryletTC, if he's up for it) have their say.

cristinasewell commented on pcdscalc/be_lens_calcs.py:
```diff
@@ -0,0 +1,939 @@
+        logger.error('The file is empyt: %s', filename)
+        return
+    with open(filename, 'rb') as lens_file:
+        sets = np.load(lens_file, allow_pickle=True)
```
I would like a bit more feedback here if anybody might have suggestions. Now that I am working on this issue https://github.com/pcdshub/pcdsdevices/issues/577 - i see that we're using this there too, and I wanted to make sure that this change is ok. 
In the old python code we used a simple config file, I am using a `.npy` file, and in the new `LensStack` class I see we're using a `.yaml` file. 
The binary file `.npy` will absolutely have to be created with the `be_lens_calcs.set_lens_set_to_file` function - I don't think it is possible to create it manually, is this something scientists will be fine with you think?

Also the `.yaml` file holds one set, but i think we want to hold more sets and be able to ask for a specific set when calling `read_lens` or `get_lens_set` - but i could be wrong here

ZLLentz replied:
Before the yaml implementation, the old python used human readable/writable files that look something like:
```
[(3, 0.0001, 1, 0.0002),
 (1, 0.0001, 1, 0.0003, 1, 0.0005),
 (2, 0.0001, 1, 0.0005)]
```
I think there is value in keeping these human read/writeable (not that I know how to interpret these...). The old python has a full directory of timestamped backup lens set files.
```
zlentz@psbuild-rhel7-01:.../xpppython_files/Be_lens_sets$ pwd
/cds/home/opr/xppopr/xpppython_files/Be_lens_sets
zlentz@psbuild-rhel7-01:.../xpppython_files/Be_lens_sets$ ls
current_set         sets_Be_2015-11-05  sets_Be_2016-02-09  sets_Be_2018-01-11  sets_Be2018-01-13  sets_Be2018-01-16  sets_Be2018-01-19  sets_Be2018-01-23  sets_Be2018-01-29
current_set_bak     sets_Be_2016-01-12  sets_Be_2018-01-09  sets_Be2018-01-11   sets_Be2018-01-14  sets_Be2018-01-17  sets_Be2018-01-21  sets_Be2018-01-24  sets_Be2018-01-31
sets_Be_2015-06-18  sets_Be_2016-01-22  sets_Be_2018-01-10  sets_Be2018-01-12   sets_Be2018-01-15  sets_Be2018-01-18  sets_Be2018-01-22  sets_Be2018-01-25
```
I'm not sure of the utility in holding multiple sets in a file. Typically the lens set file reflects the physical configuration of the lens hardware (which can be switched out between experiments). We want the python session to be able to be loaded up with a lens set file that reflects the current configuration unambiguously.

cristinasewell replied:
Thanks @ZLLentz  - I will probably have to change this back to using a different file format that's human readable. 
Regarding multiple sets, I might have not understood correctly, but this is what Eric said about them: 

`The Be lens holders can take 3 different sets that we usually set before few experiments so we only vent the relevant beamline section once. We then store these sets into a config file that we can call from the methods you redefined. Later, we save this config file to the specific experiment so users can make sure that they know which stack was used for there beamtime. Is that something we can get as well?` 

I just did not know why in the new `LensStack` class we changed to assume that there is one lens set in the file - there was probably a reason for that i am missing

ZLLentz replied:
Sorry, I think I got confused, in my head one "lens set" was the three sets together, not each of the three individually. I thought you meant having a large number of sets in the file, e.g. like 20 sets of 3, and then some selection between them.

The current implementation supporting only one set of lenses must be an error or an oversight.
